keps/sig-api-machinery/2161-apiserver-default-labels/README.md

```diff
@@ -0,0 +1,350 @@
+(i.e. “kubernetes.io/metadata.name” or some such), to allow easy selection.
+
+If this label is missing, it is added by the apiserver
+
+If the label is incorrect or changed, the apiserver fails to validate the object
+
+
+### User Stories (Optional)
+
+See https://github.com/kubernetes/enhancements/pull/2113/files for relevant user stories.
+
+### Notes/Constraints/Caveats (Optional)
+
+
+### Risks and Mitigations
+
+* Every namespace gets bigger in size. We think this is inconsequential
+* Some user may already be using that label already. This is mitigated by the 
+fact that the whole "kubernetes.io" label prefix is reserved. However, since there
+is no validation of reserved label prefixes, such namespaces will be invalidated
+post-upgrade, i.e. no further writes allowed. For this reason, we can do one alpha
+release where we log/event if we observe usage of this key that is not compatible.
+
+## Design Details
+
+This can be implemented by modifying the defaults.go and validation.go components for the namespace API, i.e., in the defaults.go file for api/core:
+
+```
+func SetDefaults_Namespace(obj *v1.Namespace) {
+  if obj.Labels == nil {
+  	obj.Labels = map[string]string{}
+  }
+  
+  if _, found := obj.Labels[“kubernetes.io/metadata.name”]; !found {
+  	obj.Labels[“kubernetes.io/metadata.name”] = obj.Name
+  }
+}
+```
+And in the validation.go, we implement the guarantee of this namespace’s value 
+being constant WRT namespace name (which is immutable).
+
+```
+func ValidateNamespace(namespace *core.Namespace) field.ErrorList {
+	// Check if namespace.Labels[“kubernetes.io/metadata.name”] == namespace.Name
+  // if not add an error
+  // ...
+  return allErrs
+```
+
+This solves not just NetworkPolicy, but all such namespace selector APIs without 
+any disruption or version skew problems.
+
+### Test Plan
+* Add unit tests that creates a namespace and checks if the namespace contains 
+the label
+* Add a test that tries to select this namespace by the label (this should 
+return also only that namespace)
+* Try to modify the reserved label and this should return an error
+* Check the WATCH operation against existing namespaces that do not have this 
+label still work properly.
+
+### Graduation Criteria
+* This feature being enabled by default at least one release.
+
+### Upgrade / Downgrade Strategy
+* Check the WATCH situation described in test plan, otherwise populate the label 
+of namespaces without it.
+* In case of downgrade, as this is just a label it will work fine. Network Policies 
+that relies on this label may fail-close, which might be acceptable.
+
+### Version Skew Strategy
+N/A
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+1.21:
+- gate disabled by default
+- if gate disabled, log or event incompatible uses of this
+- if gate enabled, enforce
+- release note
+
+1.22:
+- gate enabled by default
+- release note
```

Comment:
if we're looking for feedback from consumers, could we gain that earlier/more quickly with a namespace-labeling controller and accelerate this?

Reply:
in general most of the NetworkPolicy APIs could be prototyped in the early stages via some type of lightweight CRD or controller .  I think we should visit this monday .  cc @rikatz @abhiraut ha look the netpol operator idea ... just ... won't. .. go ... away  :)

Reply:
resolving , assuming that the controller feedback idea isnt an actual blocker for this as a feature, but more an idea we can consider